src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -0,0 +1,279 @@
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CodeActionResolveHandler(
+            ICodeFixService codeFixService,
+            ICodeRefactoringService codeRefactoringService,
+            ILspSolutionProvider solutionProvider)
+            : base(solutionProvider)
+        {
+            _codeFixService = codeFixService;
+            _codeRefactoringService = codeRefactoringService;
+        }
+
+        public override async Task<LSP.VSCodeAction> HandleRequestAsync(
+            LSP.VSCodeAction codeAction,
+            LSP.ClientCapabilities clientCapabilities,
+            string? clientName,
+            CancellationToken cancellationToken)
+        {
+            CodeActionResolveData data;
+            if (codeAction.Data is CodeActionResolveData codeActionResolveData)
+            {
+                data = codeActionResolveData;
+            }
+            else
+            {
+                data = ((JToken)codeAction.Data).ToObject<CodeActionResolveData>();
+            }
+
+            var document = SolutionProvider.GetDocument(data.CodeActionParams.TextDocument, clientName);
+            var codeActions = await CodeActionsHandler.GetCodeActionsAsync(
+                document,
+                _codeFixService,
+                _codeRefactoringService,
+                data.CodeActionParams.Range,
+                cancellationToken).ConfigureAwait(false);
+
+            if (codeActions == null || !codeActions.Any())
+            {
+                return codeAction;
+            }
+
+            var codeActionToResolve = GetCodeActionToResolve(data.DistinctTitle, codeActions);
+
+            // We didn't find a matching action, so just return the action without an edit or command.
+            if (codeActionToResolve == null)
+            {
+                return codeAction;
+            }
+
+            var operations = await codeActionToResolve.GetOperationsAsync(cancellationToken).ConfigureAwait(false);
+            if (operations.IsEmpty)
+            {
+                return codeAction;
+            }
+
+            // TO-DO:
+            // 1) We currently must execute code actions which add new documents on the server as commands,
+            // since there is no LSP support for adding documents yet. In the future, we should move these actions
+            // to execute on the client.
+            // 2) There is also a bug (same tracking item) where code actions that edit documents other than the
+            // one where the code action was invoked from do not work. We must temporarily execute these as commands
+            // as well.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+            var runAsCommand = false;
+
+            var applyChangesOperations = operations.Where(operation => operation is ApplyChangesOperation);
+            if (applyChangesOperations.Any())
```

Comment:
I wonder if this should be all, or more specific filtering.

If we have an ApplyChangesOperation + DocumentNavigation operation, then it should be safe to do this (we can't tell the client to navigate anywhere anyway, so we can't do anything but drop the document navigation operation).  But if there is an ApplyChangesOperation + some other actual modification operation, then we might get into a corrupt state b/c we applied half of the code action.

Not sure the best way to do this, thoughts @CyrusNajmabadi ?